tools/resource-gen/main.go

```diff
@@ -418,34 +407,120 @@ func main() {
 		resources = append(resources, resourceInfo)
 	}
 
-	var generatorTemplate *template.Template
+	var generatorFn GeneratorFn
 
 	switch gen {
 	case "type":
-		generatorTemplate = ResourceTemplate
+		generatorFn = TemplateGeneratorFn(ResourceTemplate)
 	case "crd":
-		generatorTemplate = CustomResourceTemplate
+		generatorFn = TemplateGeneratorFn(CustomResourceTemplate)
+	case "openapi":
+		generatorFn = openApiGenerator
 	default:
 		log.Fatalf("%s is not a valid generator option\n", gen)
 	}
 
-	outBuf := bytes.Buffer{}
-	if err := generatorTemplate.Execute(&outBuf, struct {
-		Package   string
-		Resources []ResourceInfo
-	}{
-		Package:   pkg,
-		Resources: resources,
-	}); err != nil {
-		log.Fatalf("template error: %s", err)
-	}
-
-	out, err := format.Source(outBuf.Bytes())
+	err := generatorFn(pkg, resources)
 	if err != nil {
 		log.Fatalf("%s\n", err)
 	}
+}
 
-	if _, err := os.Stdout.Write(out); err != nil {
-		log.Fatalf("%s\n", err)
+func openApiGenerator(pkg string, resources []ResourceInfo) error {
+	protoTypeToType := map[string]reflect.Type{
+		"Mesh":        reflect.TypeOf(v1alpha1.Mesh{}),
+		"MeshGateway": reflect.TypeOf(v1alpha1.MeshGateway{}),
+	}
+
+	for _, r := range resources {
+		tpe, exists := protoTypeToType[r.ResourceType]
+		println("generating", r.ResourceType)
+		if !exists {
+			continue
+		}
+		reflector := jsonschema.Reflector{
+			ExpandedStruct:            true,
+			DoNotReference:            true,
+			AllowAdditionalProperties: true,
+		}
+		err := reflector.AddGoComments("github.com/kumahq/kuma/", "api/")
+		if err != nil {
+			return err
+		}
+		schemaMap := orderedmap.New[string, *jsonschema.Schema]()
+		schemaMap.Set("type", &jsonschema.Schema{Type: "string"})
+		schemaMap.Set("name", &jsonschema.Schema{Type: "string"})
+		if r.Global {
+			schemaMap.Set("mesh", &jsonschema.Schema{Type: "string"})
+		}
+		schemaMap.Set("labels", &jsonschema.Schema{Type: "object", AdditionalProperties: &jsonschema.Schema{Type: "string"}})
+		schemaMap.Set("spec", reflector.ReflectFromType(tpe))
+
+		schema := jsonschema.Schema{
+			Type:       "object",
+			Required:   []string{"type", "name", "spec"},
+			Properties: schemaMap,
+		}
+
+		out, err := yaml.Marshal(schema)
+		if err != nil {
+			return err
+		}
+
+		outDir := path.Join("api", pkg, "v1alpha1", strings.ToLower(r.ResourceType))
+
+		// Ensure the directory exists
+		err = os.MkdirAll(outDir, 0o755)
+		if err != nil {
+			return fmt.Errorf("failed to create directory: %w", err)
+		}
+
+		err = os.WriteFile(path.Join(outDir, "schema.yaml"), out, 0o600)
+		if err != nil {
+			return err
+		}
+
+		templatePath := path.Join("tools", "openapi", "templates", "endpoints.yaml")
+		tmpl, err := template.ParseFiles(templatePath)
+		if err != nil {
+			return err
+		}
+		opts := map[string]interface{}{
+			"Package": "v1alpha1",
+			"Name":    r.ResourceType,
+			"Scope":   "Global",
```

Comment:
You should define scope:
```
scope := "Mesh"
if r.Global {
scope = "Global"
}
```
```suggestion
			"Scope":   scope,
```